### ISSUE TITLE
propagating input target information

### ISSUE THREAD

**Issue by forero** [desihub/fiberassign#96]
There is additional target information that must be propagated through `fiberassign`. Where is the best place to make this propagation? Inside `fiberassign`? In `mtl`? 
@sbailey can you give us some examples of the kind of information/columns that we should be propagating? Is it every column that can be found in the `targets`/`mtl` file?

**Comment by forero** [comment]
@sbailey  Do we know what information should be propagated? 

Related to this we have this comment from Kyle in the last internal review 

> We need to finalize the [fiberassign] outputs so that Klaus has something to test in mock observing.  We need to finalize the interface by which the fiber assignment outputs are transferred to the ICS system as part of the next field selector routine.

What is the minimal set of columns that Klaus needs?

**Comment by djschlegel** [comment]
If there are to be any changes, they should be reflected in the data model for the tile files:
  http://desidatamodel.readthedocs.io/en/latest/DESI_TARGET/fiberassign/tile.html
It looks like the current outputs do correspond to the current data model.

**Comment by sbailey** [comment]
Some details are still being worked out, but here is at least a starting set of changes to make to fiberassign output (and thus to MTL in some cases):

Columns to add:
  * [ ] SUBPRIORITY (float64), from targets -> MTL -> fiberassign
  * [ ] PETAL_LOC (int8, 0-9 = LOCATION // 1000)
  * [ ] DEVICE_LOC (int16, 0-526 = LOCATION % 1000)

Placeholder columns to add; these will come from targets -> MTL -> fiberassign but aren't in the target catalogs yet:

  * [ ] PMRA, PMDEC (float32) default 0.0 (proper motions from Gaia)
  * [ ] PMRA_IVAR, PMDEC_IVAR (float32) default 1.0 (inverse variances of PM*)
  * [ ] PARALLAX (float32) default 0.0
  * [ ] PARALLAX_IVAR (float32) default 1.0
  * [ ] Header keyword (not column) REFEPOCH = 2015.5 // Gaia Proper Motion Reference Epoch

Updates to existing columns:
  * [ ] Rename RA,DEC -> TARGET_RA, TARGET_DEC (for consistency with ICS and clarity of target vs. fiber vs. telescope...)
  * [ ] Replace BRICKNAME (S8) with BRICKID (int32)
  * [ ] Change X,Y coordinate system to Q(deg), S(mm)
    * The desimodel fiberpos.fits file has Q,S columns though the fiberpos.txt file does not; See DESI-0530 for coordinate system pictures
  * [ ] Replace X/YFOCAL_DESIGN -> DESIGN_Q/S

To be determined:
  * There will be additional targeting bits to propagate but the names are TBD.  e.g. SECONDARY_TARGET, COMM_TARGET, SV_DESI_TARGET, ...
  * We will likely want to add row-matched target catalog entries as a separate HDU; this is probably most easily done with python while adding the GFA targets HDU instead of propagating N>>1 columns through the C++ code.

I'll open a separate ticket in desitarget about adding additional columns to the select_gfa outputs.

**Comment by sbailey** [comment]
This will be part of major datamodel refactor that won't converge in time for the 18.6 release; punting to 18.9 release.

**Comment by forero** [comment]
@sbailey 

I've just opened PR #141 to solve this ticket. 

I still have a few questions on the best way to solve it. I am following https://desidatamodel.readthedocs.io/en/mockobserving/DESI_TARGET/fiberassign/tile-TILEID-FIELDNUM.html to get the list of columns to be implemented. 

Many of them are currently found in the `targets` file and not in `mtl`. Should we change how the `mtl` is computed or simply pass the `targets` file to fiberassign?

**Comment by moustakas** [comment]
A quick note since I've been staring at the latest (`targets-dr7.1-0.23.0.fits`) targets catalog recently:

> * [x] PMRA, PMDEC (float32) default 0.0 (proper motions from Gaia)
> * [x] PMRA_IVAR, PMDEC_IVAR (float32) default 1.0 (inverse variances of PM*)
> * [x] PARALLAX (float32) default 0.0
> * [x] PARALLAX_IVAR (float32) default 1.0
> * [ ] Header keyword (not column) REFEPOCH = 2015.5 // Gaia Proper Motion Reference Epoch

The latest catalog now has the checked-off columns above (although the inverse variances do not have the requested default value of 1.0) but is missing the REFEPOCH header card.

See here for more:
https://desidatamodel.readthedocs.io/en/latest/DESI_TARGET/targets.html

**Comment by moustakas** [comment]
No problem.  Also note that `PRIORITY` and `NUMOBS` are now set by `desitarget.targets.finalize`, so we need to update that part of `desitarget.mtl.make_mtl`.  (Sorry, this comment belongs elsewhere.)

**Comment by moustakas** [comment]
Fiberassign also needs to carry forward the healpixel (`HPXPIXEL`) number from the target catalog, as discussed in https://github.com/desihub/desitarget/issues/238.

**Comment by forero** [comment]
@moustakas `HPXPIXEL` will be stored in the HDU4 https://desidatamodel.readthedocs.io/en/mockobserving/DESI_TARGET/fiberassign/tile-TILEID-FIELDNUM.html#hdu4

**Comment by forero** [comment]
`TYPE`  lives in `HDU1`. What is `FRACDEV` ?

**Comment by moustakas** [comment]
It's the fraction of the model that's DEV vs EXP, documented here--
http://legacysurvey.org/dr7/catalogs/

So for example the (weighted) "size" of a general galaxy is `FRACDEV * SHAPEDEV_R + (1-FRACDEV) * SHAPEEXP_R`, and so forth.

(No need to explain here, but I'm personally confused about why some quantities are in HDU1 vs HDU4, and why many columns are repeated.)

**Comment by sbailey** [comment]
HDU4 TARGETS should be a passthrough of *all* columns from the input target file without trying to make choices of which columns are required or not.  i.e. just subselect the rows of the input target catalog that were assigned, get them in the correct sorted order, and write them out (i.e. pass through whatever you get as input, don't try to match the current as-documented datamodel; we'll update the datamodel to match).

HDU1 FIBERASSIGN is the subset of columns needed for ICS observations and spectroscopic pipeline processing, augmented by the actual fiber assignment columns.

The original hope was that size(HDU1) << side(HDU4) and not carrying around N>>1 unnecessary columns would be beneficial, while still passing forward the full input target catalog in HDU4.  I'm on the fence about whether that is working out well or not.

**Comment by forero** [comment]
Is PR #141 good enough to close this issue? HDU1 is done, but HDU4 TARGETS is not implemented at all.

**Comment by sbailey** [comment]
Please add the TARGETS HDU before closing this issue (or open a new issue just about that and close this one).  Notes:

* It doesn't have to be HDU #4, but it should have EXTNAME="TARGETS"
* It should contain *every* column from the input MTL, row-matched to the FIBERASSIGN HDU (i.e. the TARGETID column should be identical in both)
* See the `desimodel.footprint.find_points_in_tiles` pre-filtering that I added to `bin/fiberassign` so that any sorting/filtering only needs to work on small subset of the MTL per tile instead of searching the entire MTL every time.  The `mtl_tile_indices[tileid]` is a dictionary keyed by TILEID with values that are the indices of `mtl_data` for a superset of targets that may cover that tile.
* Here is a recipe for sorting one array to match the order of another, taken from another piece of code that wanted to re-order a `zcatalog` table to match a particular order of `targetids`, you'll need something similar here:
```python
#- Sort zcatalog to have TARGETID match a particular order
ii = np.argsort(np.argsort(targetids))
jj = np.argsort(zcatalog['TARGETID'])
zcatalog = zcatalog[jj[ii]]

#- That sequence of argsorts may feel like magic, so make sure we got it right
assert np.all(zcatalog['TARGETID'] == targetids)
```

**Comment by forero** [comment]
Got it.

**Comment by forero** [comment]
@geordie666 there are two fields required by this datamodel: https://desidatamodel.readthedocs.io/en/mockobserving/DESI_TARGET/fiberassign/tile-TILEID-FIELDNUM.html#hdu4
that I cannot find in `targets-dr7.1-PR372.fits`: `PSFDEPTH_W1` and `PSFDEPTH_W1`. Can those be included in a new version?

**Comment by forero** [comment]
Solved in #144.
